### ISSUE TITLE
Pylance does not handle "subset " typeddicts

### ISSUE THREAD

**Issue by Alexandre-SCHOEPP** [microsoft/pyright#10664]
## Environment data

-   Pylance version: 2025.6.2
-   OS and version: Windows 11
-   Python version (& distribution if applicable, e.g. Anaconda): 3.12.3

## Code Snippet

```python
from typing_extensions import TypedDict, NotRequired


class Config(TypedDict):
    a: str
    b: float


class UserOverride(TypedDict):
    a: NotRequired[str]


def get_default_config(user_override: UserOverride) -> Config:
    conf: Config = {
        "a": "hello world",
        "b": 1.
    }
    conf.update(user_override)  # No overloads for "update" match the provided arguments[PylancereportCallIssue]
    conf.update(**user_override)
    return conf
```

## Expected behaviour

I expect both calls to `dict.update` to type-check. The `UserOverride` `TypedDict` is a valid subset to `Config`, meaning that `Config.update(UserOverride) -> Config`

Note that Mypy handles the first call, but not the second. I opened an [issue](https://github.com/python/mypy/issues/19365) with them

**Comment by erictraut** [comment]
Pyright (the type checker that underlies pylance) is working as intended here, so this isn't a bug.

A `TypedDict` defines a specified set of keys, but dictionaries that are assignable to that TypedDict are allowed to have additional keys. For example, we could define a `TypedDict` named `Other` as follows, and a value that is assignable to `Other` is also assignable to `UserOverride` in your example.

```python
class Other(TypedDict):
    a: NotRequired[str]
    b: str

f = Other(b="b")
u: UserOverride = f
```

If a value of type `UserOverride` were used as an argument to `Config.update()`, it could overwrite the value of `b` with a value that is not a `float`. This is why it's unsafe to allow `conf.update(user_override)`.

To make this type safe, we'd need to guarantee that any value compatible with `UserOverride` does not have a `b`. There is a way to do this with a new type system feature described in [PEP 728](https://peps.python.org/pep-0728/). It's called a "closed TypedDict", and here's how one would specify that `UserOverride` is "closed".

```python
class UserOverride(TypedDict, closed=True):
    a: NotRequired[str]
```

If you make [this change](https://pyright-play.net/?enableExperimentalFeatures=true&reportMissingModuleSource=false&code=GYJw9gtgBALgngBwJYDsDmB9ApgDxllAZyTCKiQgTBBigBVEsATAESQGMYAaKAOTBgAlLAEcArkhDMAULPYAbAIaFCUAMKlgSNAAoGCZm04BKAFzSolqItNRCMEBasAjW8HlhFMWdIXLVAKqEWCAA8gBuISBITFh6jKwc3FAKYMFMALx0IGJYZk6WNnwCwuKSzADa9iAAuj6xwFBoWDAYDYpi8q3smto6YsEgGGCRINGxtkEhEVExeVAAtAB86r1o5lYpmrYaKFpoUBlQAN4Fm1AARIoXthcAFljyHlAA7tTyTBdcZ5sXzjdQACMADozgBfM49PbAsQIJheOIDELDUbjPKQzQwuEInQAKlxSKGI1msWMZykMDEIBQWz20iAA) in your code, it will type check without error.

PEP 728 is still in the draft phase, but pyright has provisional support for it. I recommend not using this functionality in production code because the design could still change or be rejected, but if you want to play around with it, you can set `enableExperimentalFeatures` to `true` in your pyright config file. 

The reason that pyright accepts the `conf.update(**user_override)` statement has to do with an under-specified part of the Python typing spec regarding the unpack operator when applied to arguments within a call expressions. If you're interested in this topic, refer to [this thread in the Python typing forum](https://discuss.python.org/t/standardizing-argument-unpacking-behaviors-for-calls/80747) where we discuss various options for standardizing the behavior.
